scene/animation/skeleton_ik.h

```diff
@@ -0,0 +1,202 @@
+				pb(NULL),
+				length(0) {}
+
+		ChainItem *find_child(const BoneId p_bone_id);
+		ChainItem *add_child(const BoneId p_bone_id);
+	};
+
+	struct ChainTip {
+		ChainItem *chain_item;
+		const EndEffector *end_effector;
+
+		ChainTip() :
+				chain_item(NULL),
+				end_effector(NULL) {}
+
+		ChainTip(ChainItem *p_chain_item, const EndEffector *p_end_effector) :
+				chain_item(p_chain_item),
+				end_effector(p_end_effector) {}
+
+		ChainTip(const ChainTip &p_other_ct) :
+				chain_item(p_other_ct.chain_item),
+				end_effector(p_other_ct.end_effector) {}
+	};
+
+	struct Chain {
+		ChainItem chain_root;
+		Vector<ChainTip> tips;
+	};
+
+public:
+	struct Task : public RID_Data {
+		RID self;
+		Skeleton *skeleton;
+
+		Chain chain;
+
+		// Settings
+		real_t min_distance;
+		int max_iterations;
+
+		// Bone data
+		BoneId root_bone;
+		Vector<EndEffector> end_effectors;
+
+		Transform goal_global_transform;
+
+		Task() :
+				skeleton(NULL),
+				min_distance(0.01),
+				max_iterations(10),
+				root_bone(-1) {}
+	};
+
+private:
+	/// Init a chain that starts from the root to tip
+	static void build_chain(Task *p_task, bool p_force_simple_chain = true);
+
+	static void update_chain(const Skeleton *p_sk, ChainItem *p_chain_item);
+
+	/// Special solvers that solve only chains with one end effector
+	static void solve_simple_backwards(Chain &r_chain);
+	static void solve_simple_forwards(Chain &r_chain);
+
+public:
+	static Task *create_simple_task(Skeleton *p_sk, BoneId root_bone, BoneId tip_bone, const Transform &goal_transform);
+	static void free_task(Task *p_task);
+	// The goal of chain should be always in local space
+	static void set_goal(Task *p_task, const Transform &p_goal);
+	static void make_goal(Task *p_task, const Transform &p_inverse_transf, real_t blending_delta);
+	static void solve(Task *p_task, real_t blending_delta);
+};
+
+class SkeletonIK : public Node {
+	GDCLASS(SkeletonIK, Node);
+
+	String root_bone;
+	String tip_bone;
+	real_t interpolation;
+	Transform target;
+	NodePath target_node_path_override;
+
+	real_t min_distance;
+	int max_iterations;
+
+	Skeleton *skeleton;
+	Spatial *target_node_override;
+	FabrikInverseKinematic::Task *task;
+	bool process_solve;
+
+protected:
+	virtual void
+	_validate_property(PropertyInfo &property) const;
+
+	static void _bind_methods();
+	virtual void _notification(int p_notification);
+
+public:
+	SkeletonIK();
+	virtual ~SkeletonIK();
+
+	void set_root_bone(const String &p_root_bone);
+	const String &get_root_bone() const;
+
+	void set_tip_bone(const String &p_tip_bone);
+	const String &get_tip_bone() const;
+
+	void set_interpolation(real_t p_interpolation);
+	real_t get_interpolation() const;
+
+	void set_target(const Transform &p_target);
+	const Transform &get_target() const;
+
+	void set_target_node_override(const NodePath &p_node);
+	NodePath get_target_node_override();
+
+	void set_min_distance(real_t p_min_distance);
+	real_t get_min_distance() const { return min_distance; }
```

Comment:
I dont quite understand what this is for

Reply:
This is used to control the precision of IK, it tells the minimum distance between target and tip before consider the process done.